### DISCUSSION TITLE
TypeAdapter and mypy compatibility

### DISCUSSION THREAD

**Discussion by canassa** [pydantic/pydantic#7094] [Question] [answered]
Hello, 

I've implemented a function utilizing Pydantic V2 which leverages TypeAdapter and a discriminated union to convert a given object into one of the possible BaseModel types defined in my project. Here's the function:

```python
from typing import Annotated

from pydantic import Field, TypeAdapter

from myproject import AllMessages

Payload = Annotated[AllMessages, Field(discriminator="message_type")]
shape_adapter = TypeAdapter(Payload)


def parse_obj(obj: dict) -> AllMessages:
    return shape_adapter.validate_python(obj)  # type: ignore
```

While the function operates correctly, I'm encountering a mypy typing issue that I've had to bypass using # type: ignore. Here's the error I'm receiving:

```
_parser.py:12: error: Incompatible return value type (got "<typing special form>", expected "Msg1 | Msg2 | Msg3")  [return-value]
```

Does anyone know how to get this to work with mypy?

**Comment by kwint** [comment]
I have the exact same issue, but I think this is more of a mypy issue. Looking at pydantic's code I don't see anything wrong

**Comment by Viicos** [comment]
It seems mypy doesn't like playing with special forms:

```python
Payload: TypeAlias = A | B
reveal_type(Payload)  # note: Revealed type is "typing._SpecialForm"
```

While this is true from the Python perspective (`Union` is an instance of `_SpecialForm`), I don't think it helps in this case

**Comment by cj81499** [comment]
I have the same issue in a work project of mine.

See also: https://github.com/pydantic/pydantic/issues/5580#issuecomment-1717917991

**Reply by Viicos** [reply]
Unfortunately Pydantic can't really do anything here, it is making use of specific typing objects called special forms (such as `Annotated`, `Union`, `Optional`, etc). Using them as _values_ (e.g. `TypeAdapter(Union[int, str])`) instead as _annotations_ (e.g. `value: Union[int, str]`) isn't specified, and both pyright (since 1.1.346) and mypy don't support them.

A concept called "type forms" might solve this issue (see https://github.com/python/mypy/issues/9773), but nothing has been proposed yet.

**Reply by exhumea** [reply]
This `mypy` issue is also happening for something as simple as (tested with mypy==1.12.0, pydantic==2.9.2):

```python
import typing
from collections.abc import Sequence

from pydantic import TypeAdapter

ADAPTER = TypeAdapter(Sequence[int])  # list[int] would be OK though (but not read-only)
xs = ADAPTER.validate_python([1, "2", 3.0])  # [1, 2, 3]

typing.reveal_type(ADAPTER)  # pydantic.type_adapter.TypeAdapter[Any]
typing.reveal_type(xs)  # Any
```

`pyright==1.1.385` performs well though.

**Comment by AlexanderPodorov** [comment]
For that's worth I've found a workaround, e.g.:
`RootModel[int | str].model_validate('abc').root`.

**Comment by Viicos** [comment, marked as answer]
https://github.com/pydantic/pydantic/pull/13459 adds support for [`TypeForm`](https://docs.python.org/3.15/library/typing.html#typing.TypeForm) introduced by [PEP 747](https://peps.python.org/pep-0747/), which fixes the encountered type checking issue.

This will be available in the next release.
